port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,411 @@
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/**
+ * @file
+ * @ingroup Port
+ * @brief Stack backtracing support
+ */
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include "omrport.h"
+#include "omrportpriv.h"
+#include "omrsignal_context.h"
+
+#include <dlfcn.h>
+#include <execinfo.h>
+#include <stdlib.h>
+#include <string.h>
+#define UNW_LOCAL_ONLY
+#include <libunwind.h>
+
+#include "omrintrospect.h"
+
+uintptr_t protectedBacktrace(OMRPortLibrary *port, void *arg);
+uintptr_t backtrace_sigprotect(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, void **address_array, const int capacity);
+
+typedef struct frameData {
+	void **address_array;
+	unsigned int capacity;
+} frameData;
+
+
+/**
+ * Handles segfaults where either an invalid address is called or a signal was received in kernel mode.
+ * Both cases leave the ip register address invalid and libunwind fails to iterate past _sigtramp. This
+ * function fetches previous frame information and restores correct register values to the unwind cursor.
+ * It also records the stack frame addresses of the remaining frames.
+ *
+ * @param[in] cursor keeps track of stack unwind state
+ * @param[out] array contains the backtrace addresses to be returned
+ * @param[in] size size of the array
+ * @param[in] i number of addresses currently recorded in the array
+ *
+ * @return total number of addresses recorded in the array
+ */
+static uintptr_t
+handle_sigtramp(unw_cursor_t *cursor, void **array, const int size, int i)
+{
+	unw_word_t prevFrame = 0;
+	unw_word_t ip = 0;
+	ucontext_t *uc = NULL;
+	OMRUnixSignalInfo info;
+	const char *name = NULL;
+	void *reg = 0;
+
+#if defined(OMR_ARCH_X86)
+	/* ucontext for frame prior to _sigtramp is stored in rbx. Fetch previous frame info from it.
+	 * See https://opensource.apple.com/source/Libc/Libc-498/x86_64/sys/_sigtramp.s.auto.html
+	 */
+	unw_get_reg(cursor, UNW_X86_64_RBX, &prevFrame);
+	uc = (ucontext_t *)prevFrame;
+	info.platformSignalInfo.context = uc;
+
+	/* Copy the necessary machine regs to cursor so we can iterate the previous frame.
+	 * See https://opensource.apple.com/source/libunwind/libunwind-35.4/src/unw_getcontext.s.auto.html
+	 */
+	infoForControl(NULL, &info, 6, &name, &reg);
+	unw_set_reg(cursor, UNW_X86_64_RBP, *(unw_word_t *)reg);
```

Comment:
Shouldn't this (and line 91) be a use of `unw_get_reg()`? I don't think we want to _modify_ the state of any stack frame.

Reply:
We modify the current stack frame (`_sigtramp`) because libunwind doesn't know how to iterate past it (according to a source on the internet) so we change these registers to the frame before `_sigtramp` and libunwind can continue iterating. The `cursor` points to inside a `unw_context_t` structure and I assume it looks at the bp, sp, and ip to determine which frame to go to next.

Reply:
`unw_set_reg` doesn't modify the actual stack frames, just the `unw_context_t` structure which stores a copy of the stack information.